scipy/optimize/least_squares.py

```diff
@@ -0,0 +1,657 @@
+
+    >>> def fun_broyden(x):
+    ...     f = (3 - x) * x + 1
+    ...     f[1:] -= x[:-1]
+    ...     f[:-1] -= 2 * x[1:]
+    ...     return f
+
+    The corresponding Jacobian matrix is sparse. We tell the algorithm to
+    estimate it by finite differences and provide the sparsity structure of
+    Jacobian to significantly speed up this process.
+
+    >>> from scipy.sparse import lil_matrix
+    >>> def sparsity_broyden(n):
+    ...     sparsity = lil_matrix((n, n), dtype=int)
+    ...     i = np.arange(n)
+    ...     sparsity[i, i] = 1
+    ...     i = np.arange(1, n)
+    ...     sparsity[i, i - 1] = 1
+    ...     i = np.arange(n - 1)
+    ...     sparsity[i, i + 1] = 1
+    ...     return sparsity
+    ...
+    >>> n = 100000
+    >>> x0_broyden = -np.ones(n)
+    ...
+    >>> res_3 = least_squares(fun_broyden, x0_broyden,
+    ...                       jac_sparsity=sparsity_broyden(n))
+    >>> res_3.cost
+    4.5687161966109073e-23
+    >>> res_3.optimality
+    1.1650454296851518e-11
+    """
+    if method not in ['trf', 'dogbox', 'lm']:
+        raise ValueError("`method` must be 'trf', 'dogbox' or 'lm'.")
+
+    if tr_solver not in [None, 'exact', 'lsmr']:
+        raise ValueError("'tr_solver' must be None, 'exact' or 'lsmr'.")
+
+    if verbose not in [0, 1, 2]:
+        raise ValueError("`verbose` must be in [0, 1, 2].")
+
+    if len(bounds) != 2:
+        raise ValueError("`bounds` must contain 2 elements.")
+
+    x0 = np.atleast_1d(x0).astype(float)
+
+    if x0.ndim > 1:
+        raise ValueError("`x0` must have at most 1 dimension.")
+
+    lb, ub = prepare_bounds(bounds, x0)
+
+    if lb.shape != x0.shape or ub.shape != x0.shape:
+        raise ValueError("Inconsistent shapes between bounds and `x0`.")
+
+    if np.any(lb >= ub):
+        raise ValueError("Each lower bound mush be strictly less than each "
+                         "upper bound.")
+
+    if method == 'lm' and not np.all((lb == -np.inf) & (ub == np.inf)):
+        raise ValueError("Method 'lm' doesn't support bounds.")
+
+    if jac not in ['2-point', '3-point'] and not callable(jac):
+        raise ValueError("`jac` must be '2-point', '3-point' or callable.")
+
+    scaling = check_scaling(scaling, x0)
+
+    ftol, xtol, gtol = check_tolerance(ftol, xtol, gtol)
+
+    if not in_bounds(x0, lb, ub):
+        raise ValueError("`x0` is infeasible.")
+
+    def fun_wrapped(x):
+        return np.atleast_1d(fun(x, *args, **kwargs))
+
+    f0 = fun_wrapped(x0)
+    if f0.ndim != 1:
+        raise RuntimeError("`fun` must return at most 1-d array_like.")
+
+    n = x0.size
+    m = f0.size
+    if method == 'lm' and m < n:
+        raise ValueError("Method 'lm' doesn't work when the number of "
+                         "residuals is less than the number of variables.")
+
+    if jac in ['2-point', '3-point']:
+        if jac_sparsity is not None:
+            if method == 'lm':
+                warn("`jac_sparsity` is ignored for method='lm', dense "
```

Comment:
LM and sparsity do not play together, this can never be a useful thing. 
Generally, with this many mysterious switches I think it makes sense to weed out obviously wrong things.